e3nn/spherical_tensor.py

```diff
@@ -90,3 +96,283 @@ def spherical_harmonics_coeff_to_sphere(coeff, alpha, beta):
+
+        if sum_points:
+            signal = signal.sum(0)
+        new_cls = cls(signal, Rs)
+        new_cls.radial_model = radial_model
+        return new_cls
+
+    def sph_norm(self):
+        Rs = self.Rs
+        signal = self.signal
+        n_mul = sum([mul for mul, L in Rs])
+        # Keep shape after Rs the same
+        norms = torch.zeros(n_mul, *signal.shape[1:])
+        sig_index = 0
+        norm_index = 0
+        for mul, L in Rs:
+            for _ in range(mul):
+                norms[norm_index] = signal[sig_index: sig_index + (2 * L + 1)].norm(2, 0)
+                norm_index += 1
+                sig_index += 2 * L + 1
+        return norms
+
+    def signal_on_sphere(self, which_mul=None, n=100, radius=True):
+        n_mul = sum([mul for mul, L in self.Rs])
+        if which_mul:
+            if len(which_mul) != n_mul:
+                raise ValueError("which_mul and number of multiplicities is not equal.")
+        else:
+            which_mul = [1 for i in range(n_mul)]
+
+        # Need to handle if signal is featurized
+        x, y, z = (None, None, None)
+        Ys = []
+        for mul, L in self.Rs:
+            # Using cache-able function
+            x, y, z, Y = spherical_harmonics_on_grid(L, n)
+            Ys += [Y] * mul
+
+        f = self.signal.unsqueeze(1).unsqueeze(2) * torch.cat(Ys, dim=0)
+        f = f.sum(0)
+        return x, y, z, f
+
+    def plot(self, which_mul=None, n=100, radius=True, center=None, relu=True):
+        """
+        surface = self.plot()
+        fig = go.Figure(data=[surface])
+        fig.show()
+        """
+        import plotly.graph_objs as go
+
+        x, y, z, f = self.signal_on_sphere(which_mul, n, radius)
+        f = f.relu() if relu else f
+
+        if radius:
+            r = f.abs()
+            x = x * r
+            y = y * r
+            z = z * r
+
+        if center is not None:
+            x = x + center[0]
+            y = y + center[1]
+            z = z + center[2]
+
+        return go.Surface(x=x.numpy(), y=y.numpy(), z=z.numpy(), surfacecolor=f.numpy())
+
+    def plot_with_radial(self, box_length, center=None,
+                         sh=rsh.spherical_harmonics_xyz, n=30,
+                         radial_model=None, relu=True):
+        muls, _Ls = zip(*self.Rs)
+        # We assume radial functions are repeated across L's
+        assert len(set(muls)) == 1
+        num_L = len(self.Rs)
+        if radial_model is None:
+            radial_model = self.radial_model
+
+        def new_radial(x):
+            return radial_model(x).repeat(1, num_L)  # Repeat along filter dim
+        r, f = plot_data_on_grid(box_length, new_radial, self.Rs, sh=sh, n=n)
+        # Multiply coefficients
+        f = torch.einsum('xd,d->x', f, self.signal)
+        f = f.relu() if relu else f
+
+        if center is not None:
+            r += center.unsqueeze(0)
+
+        return r, f
+
+    def wigner_D_on_grid(self, n):
```

Comment:
`wigner_D_on_grid` and `cross_correlation` can be deleted, I have never used them.

Reply:
Delete then